Avoid text cursor from being applied if `-webkit-user-select: none` is set

AlexsanderDamaceno opened the pull request:
#### 4264524b4c6097fea9b173afec3f4c00cd57b213
<pre>
Avoid text cursor from being applied if `-webkit-user-select: none` is set
<a href="https://bugs.webkit.org/show_bug.cgi?id=244944">https://bugs.webkit.org/show_bug.cgi?id=244944</a>

Reviewed by Tim Nguyen.

The cursor was incorrectly changing to a text pointer even when `-webkit-user-select: none` was applied.
This is fixed by verifying the `-webkit-user-select` in the event handler and preventing the text pointer from being selected when `-webkit-user-select: none` is set.

* Source/WebCore/css/CSSProperties.json:
* LayoutTests/editing/selection/user-select-none-02-expected.txt: Added.
* LayoutTests/editing/selection/user-select-none-02.html: Added.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::selectCursor):

Canonical link: <a href="https://commits.webkit.org/289793@main">https://commits.webkit.org/289793@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f42edc1505a4fc2c2a6fbe445400e42f06a8d502

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87972 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7488 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42364 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92875 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38722 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90023 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7869 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15708 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67905 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25647 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90974 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6034 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79592 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48274 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5810 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34000 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37829 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76209 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34881 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94739 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15140 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11162 "Found 1 new test failure: ipc/decode-object-array-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76760 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15395 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75448 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75998 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18697 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20391 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18807 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8144 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15158 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20459 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14900 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18345 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16682 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -8834,7 +8834,10 @@
+                "parser-grammar": "<<values>>",
+                 "aliases": [
+                    "user-select"
+                ]
```
Unprefixing user-select is a whole other issue that requires more changes. Please check https://bugs.webkit.org/showdependencytree.cgi?id=208677&hide_resolved=1 for a list of bugs to fix before unprefixing user-select

nt1m replied:
You're free to fix this bug separately for `-webkit-user-select` specifically though.

AlexsanderDamaceno replied:
Could we keep it for now as aliases for ` -webkit-user-select ` ? then later we could set  `-webkit-user-select `as alias for user-select?

nt1m replied:
I ended up getting web compatibility issues, when I did that 3 years ago, simply because the `user-select` requires more changes.

nt1m commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1669,10 +1669,10 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
                 return iBeam;
 
-        if ((editable || (renderer && renderer->isRenderText() && node->canStartSelection())) && !inResizer && !result.scrollbar())
+        if ((editable || (renderer && renderer->isRenderText() && node->canStartSelection())) && !inResizer && !result.scrollbar() && !(renderer->style().usedUserSelect()  == UserSelect::None))
```
This should indeed fix the cursor, but maybe consider just disabling live text analysis from happening altogether? Not sure what @whsieh thinks here, I don't have a strong opinion.

AlexsanderDamaceno replied:
@nt1m  how could disable the live text analysis?

AlexsanderDamaceno replied:
@nt1m  I think the live text analysis you're referring to is about detecting text in an image, right? The bug was filed with the assumption that the problem was related to having an image element, but in reality, the bug occurs in general and does not require an image element, so in this case I think disabling live text analysis would no solve the problem.

whsieh commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1669,10 +1669,10 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
 #endif
             && frame->selection().isCaretOrRange()
-            && !m_capturingMouseEventsElement)
+            && !m_capturingMouseEventsElement &&  !(renderer->style().usedUserSelect()  == UserSelect::None))
```
Nit - extra spaces. I'd also use `!=` rather than `!(… == …)`

```suggestion
            && !m_capturingMouseEventsElement && renderer->style().usedUserSelect()  != UserSelect::None)
```

whsieh commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1669,10 +1669,10 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
                 return iBeam;
 
-        if ((editable || (renderer && renderer->isRenderText() && node->canStartSelection())) && !inResizer && !result.scrollbar())
+        if ((editable || (renderer && renderer->isRenderText() && node->canStartSelection())) && !inResizer && !result.scrollbar() && !(renderer->style().usedUserSelect()  == UserSelect::None))
```
```suggestion
        if ((editable || (renderer && renderer->isRenderText() && node->canStartSelection())) && !inResizer && !result.scrollbar() && renderer->style().usedUserSelect() != UserSelect::None)
```

Ahmad-S792 commented on LayoutTests/editing/selection/user-select-none-02.html:
```diff
@@ -0,0 +1,31 @@
+}
+</style>
+</head>
+<script src="../../resources/js-test-pre.js"></script>
```
`js-test.js` is defacto going forward (or `testharness.js` similar to WPT). So use it instead of `js-test-pre.js`.

AlexsanderDamaceno committed:
Avoid text cursor from being applied if `-webkit-user-select: none` is set
https://bugs.webkit.org/show_bug.cgi?id=244944

Reviewed by Tim Nguyen.

The cursor was incorrectly changing to a text pointer even when `-webkit-user-select: none` was applied.
This is fixed by verifying the `-webkit-user-select` in the event handler and preventing the text pointer from being selected when `-webkit-user-select: none` is set.

* Source/WebCore/css/CSSProperties.json:
* LayoutTests/editing/selection/user-select-none-02-expected.txt: Added.
* LayoutTests/editing/selection/user-select-none-02.html: Added.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::selectCursor):

Canonical link: https://commits.webkit.org/289793@main
